src/server/frontend_wayland/text_input_v1.cpp

```diff
@@ -0,0 +1,421 @@
+
+#undef WAYLAND_TO_MIR_HINT
+}
+
+auto wayland_to_mir_content_purpose(uint32_t purpose) -> ms::TextInputContentPurpose
+{
+    switch (purpose)
+    {
+        case mw::TextInputV1::ContentPurpose::normal:
+            return ms::TextInputContentPurpose::normal;
+
+        case mw::TextInputV1::ContentPurpose::alpha:
+            return ms::TextInputContentPurpose::alpha;
+
+        case mw::TextInputV1::ContentPurpose::digits:
+            return ms::TextInputContentPurpose::digits;
+
+        case mw::TextInputV1::ContentPurpose::number:
+            return ms::TextInputContentPurpose::number;
+
+        case mw::TextInputV1::ContentPurpose::phone:
+            return ms::TextInputContentPurpose::phone;
+
+        case mw::TextInputV1::ContentPurpose::url:
+            return ms::TextInputContentPurpose::url;
+
+        case mw::TextInputV1::ContentPurpose::email:
+            return ms::TextInputContentPurpose::email;
+
+        case mw::TextInputV1::ContentPurpose::name:
+            return ms::TextInputContentPurpose::name;
+
+        case mw::TextInputV1::ContentPurpose::password:
+            return ms::TextInputContentPurpose::password;
+
+        case mw::TextInputV1::ContentPurpose::date:
+            return ms::TextInputContentPurpose::date;
+
+        case mw::TextInputV1::ContentPurpose::time:
+            return ms::TextInputContentPurpose::time;
+
+        case mw::TextInputV1::ContentPurpose::datetime:
+            return ms::TextInputContentPurpose::datetime;
+
+        case mw::TextInputV1::ContentPurpose::terminal:
+            return ms::TextInputContentPurpose::terminal;
+
+        default:
+            BOOST_THROW_EXCEPTION(std::invalid_argument{"Invalid text content purpose " + std::to_string(purpose)});
+    }
+}
+
+struct TextInputV1Ctx
+{
+    std::shared_ptr<mir::Executor> const wayland_executor;
+    std::shared_ptr<ms::TextInputHub> const text_input_hub;
+};
+
+class TextInputManagerV1Global
+    : public mw::TextInputManagerV1::Global
+{
+public:
+    TextInputManagerV1Global(wl_display* display, std::shared_ptr<TextInputV1Ctx> ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<TextInputV1Ctx> const ctx;
+};
+
+class TextInputManagerV1
+    : public mw::TextInputManagerV1
+{
+public:
+    TextInputManagerV1(wl_resource* resource, std::shared_ptr<TextInputV1Ctx> ctx);
+
+private:
+    void create_text_input(wl_resource* id) override;
+
+    std::shared_ptr<TextInputV1Ctx> const ctx;
+};
+
+class TextInputV1
+    : public mw::TextInputV1,
+      private mf::WlSeat::FocusListener
+{
+public:
+    TextInputV1(wl_resource* resource, std::shared_ptr<TextInputV1Ctx> const& ctx);
+    ~TextInputV1();
+
+private:
+    struct Handler : ms::TextInputChangeHandler
+    {
+        Handler(TextInputV1* const text_input, std::shared_ptr<mir::Executor> const wayland_executor)
+            : text_input{text_input},
+              wayland_executor{wayland_executor}
```

Comment:
Here, we have problematic top-level const, which prevents `wayland_executor{std::move(wayland_executor)}`